### PR TITLE
Limit client connections connecting to gnmi server

### PR DESCRIPTION
**Pull request by zbud-msft**
#### Why I did it

Incident in which client kept creating connections resulting in memory utilization alert

#### How I did it

Implemented threshold feature such that there is a total amount of 100 connections allowed and no more.

#### How to verify it

End to end test/UT

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: gnmi_server/client_subscribe.go [hunk 1]
```diff
@@ -130,6 +139,11 @@ func (c *Client) Run(stream gnmipb.GNMI_SubscribeServer) (err error) {
 	if err != nil {
 		return grpc.Errorf(codes.NotFound, "Invalid subscription path: %v %q", err, query)
 	}
+
+	if connectionKey, valid = validateConnectionRequest(c.addr, c.threshold, query.String()); !valid {
+		return grpc.Errorf(codes.ResourceExhausted, "Server connections are at capacity.")
```
Comment (ayelrod) [inline gnmi_server/client_subscribe.go:144]: I believe that the UNAVAILABLE code makes more sense here to indicate a transient error.

File: telemetry/telemetry.go [hunk 2]
```diff
@@ -57,6 +58,12 @@ func main() {
 		log.Errorf("port must be > 0.")
 		return
 	}
+
+	switch {
+	case *threshold <= 0:
```
Comment (ayelrod) [inline telemetry/telemetry.go:63]: Could you add a way to turn this threshold off? For example: if threshold == 0, then there is no limit on the number of connections.

File: gnmi_server/client_subscribe.go [hunk 3]
```diff
@@ -31,13 +30,17 @@ type Client struct {
 	w     sync.WaitGroup
 	fatal bool
 	logLevel   int
+	threshold  int
```
[the remark this answers is not in the document]
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go:33]: No, I will instantiate connection manager instance with threshold and pass to client_subscribe, therefore client will not need to store value.

File: gnmi_server/client_subscribe.go (hunk 1 above)
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go:144]: Agree, will change.

File: telemetry/telemetry.go (hunk 2 above)
Comment (zbud-msft) [reply telemetry/telemetry.go:63]: Good idea, I will make "-1" signify no threshold limit.

**Comment by ganglyu** [comment]
Unit test is broken?

File: gnmi_server/client_subscribe.go [hunk 4]
```diff
@@ -174,6 +188,8 @@ func (c *Client) Run(stream gnmipb.GNMI_SubscribeServer) (err error) {
 	c.Close()
 	// Wait until all child go routines exited
 	c.w.Wait()
+	// delete connectionkey under ip/connections
+	connectionManager.Remove(connectionKey)
```
Comment (ganglyu) [inline gnmi_server/client_subscribe.go:192]: There are some other "return grpc.Errorf(...)" in this function, do we need to remove connection for them?

File: gnmi_server/client_subscribe.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ganglyu) [reply gnmi_server/client_subscribe.go:33]: We don't need this threshold per client, global variable is enough.

File: gnmi_server/connection_manager.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	defer cm.mu.Unlock()
+	if len(cm.connections) >= cm.threshold {
+		log.V(1).Infof("Cannot add another client connection as threshold is already at limit")
+		return "", false
+	}
+	key := createKey(addr, query)
+	log.V(1).Infof("Adding client connection: %s", key)
+	cm.connections[key] = struct{}{}
+	log.V(1).Infof("Current number of existing connections: %d", len(cm.connections))
+	storeKeyRedis(key)
```
Comment (ganglyu) [inline gnmi_server/connection_manager.go:41]: Why do you store the key in redis?

File: gnmi_server/client_subscribe.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go]: Done

File: gnmi_server/client_subscribe.go (hunk 1 above)
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go]: Done

File: telemetry/telemetry.go (hunk 2 above)
Comment (zbud-msft) [reply telemetry/telemetry.go]: Done

File: gnmi_server/client_subscribe.go (hunk 4 above)
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go:196]: Yes, good catch. Done

File: gnmi_server/connection_manager.go (hunk 5 above)
Comment (zbud-msft) [reply gnmi_server/connection_manager.go:56]: For debugging purposes

File: gnmi_server/connection_manager.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+type ConnectionManager struct {
+	connections  map[string]struct{}
+	mu           sync.Mutex
+	threshold    int
+}
+
+func (cm *ConnectionManager) PrepareRedis() {
+        ns := sdcfg.GetDbDefaultNamespace()
+	rclient = redis.NewClient(&redis.Options{
```
Comment (qiluo-msft) [inline gnmi_server/connection_manager.go:26]: >	 [](http://example.com/codeflow?start=0&length=1)

Wrong indentation due to mixing tabs/spaces.

File: gnmi_server/server_test.go [hunk 7]
```diff
@@ -107,7 +107,7 @@ func createServer(t *testing.T, port int64) *Server {
 	}
 
 	opts := []grpc.ServerOption{grpc.Creds(credentials.NewTLS(tlsCfg))}
-	cfg := &Config{Port: port, EnableTranslibWrite: true, EnableNativeWrite: true}
+	cfg := &Config{Port: port, EnableTranslibWrite: true, EnableNativeWrite: true, Threshold: 100}
```
Comment (qiluo-msft) [inline gnmi_server/server_test.go:110]: >100 [](http://example.com/codeflow?start=91&length=3)

If there is no ConfigDB entry, you will use default value (100). Maybe there is use case that user want to have static default value changed to other values. Is it better to move this default value to a config file inside telemetry container?

File: gnmi_server/client_subscribe.go [hunk 8]
```diff
@@ -146,6 +162,7 @@ func (c *Client) Run(stream gnmipb.GNMI_SubscribeServer) (err error) {
 	}
 
 	if err != nil {
+		connectionManager.Remove(connectionKey)
```
Comment (ganglyu) [inline gnmi_server/client_subscribe.go:165]: How about defer?
defer connectionManager.Remove(connectionKey);

File: gnmi_server/server_test.go [hunk 9]
```diff
@@ -134,6 +134,25 @@ func createReadServer(t *testing.T, port int64) *Server {
...
+	if err != nil {
+		t.Errorf("could not load server key pair: %s", err)
+	}
+	tlsCfg := &tls.Config{
+		ClientAuth:   tls.RequestClientCert,
+		Certificates: []tls.Certificate{certificate},
+	}
+
+	opts := []grpc.ServerOption{grpc.Creds(credentials.NewTLS(tlsCfg))}
+	cfg := &Config{Port: port, EnableTranslibWrite: true,  Threshold: -1}
```
Comment (ganglyu) [inline gnmi_server/server_test.go:148]: This test will reject any subscribe request.
I'd like to use bigger threshold for test, for example, GNMI accepts the first request, and reject the second request.

File: telemetry/telemetry.go [hunk 10]
```diff
@@ -57,6 +58,12 @@ func main() {
 		log.Errorf("port must be > 0.")
 		return
 	}
+
+	switch {
+	case *threshold < 0:
+		log.Errorf("threshold must be > 0.")
```
Comment (ganglyu) [inline telemetry/telemetry.go:64]: But createRejectServer uses -1 as threshold?

File: gnmi_server/client_subscribe.go (hunk 8 above)
Comment (zbud-msft) [reply gnmi_server/client_subscribe.go]: Done

File: gnmi_server/connection_manager.go (hunk 6 above)
Comment (zbud-msft) [reply gnmi_server/connection_manager.go:26]: Fixed.

File: gnmi_server/server_test.go (hunk 9 above)
Comment (zbud-msft) [reply gnmi_server/server_test.go]: Set threshold to 2, so will accept first 2 subscribe requests and rejects the third.

File: telemetry/telemetry.go (hunk 10 above)
Comment (zbud-msft) [reply telemetry/telemetry.go:64]: Test now uses 2 as threshold.

**Review by qiluo-msft — APPROVED** [review]

File: gnmi_server/server_test.go [hunk 11]
```diff
@@ -2791,7 +2812,7 @@ func TestCPUUtilization(t *testing.T) {
 		    gotNoti = append(gotNoti, nn)
                 } else {
                     gotNoti = append(gotNoti, n)
-	        }
+	    }
```
Comment (ganglyu) [inline gnmi_server/server_test.go:2815]: Tab?

Comment (zbud-msft) [reply gnmi_server/server_test.go:2815]: Fixed

File: gnmi_server/connection_manager.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+			continue
+		}
+		connectionKey += matches[i][1] // index 1 contains the value we need
+		connectionKey += "|"
+	}
+	connectionKey += time.Now().UTC().Format(time.RFC3339)
+	return connectionKey
+}
+
+func storeKeyRedis(key string) {
```
[the remark this answers is not in the document]
Comment (zbud-msft) [reply gnmi_server/connection_manager.go:94]: Added

**Review by ganglyu — APPROVED** [review]

File: telemetry/telemetry.go (hunk 10 above)
Comment (ayelrod) [inline telemetry/telemetry.go:64]: nit: "threshold must be >= 0"

**Review by ayelrod — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Review by ganglyu — APPROVED** [review]

**Pull request merged**